File: 2016/20/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	input := strings.Split(string(data), "\n")
	fmt.Println(solve(input))
}

func solve(input []string) (int, int) {
	ranges := [][2]int{}
	for _, line := range input {
		nums := strings.Split(line, "-")
		n1, _ := strconv.Atoi(nums[0])
		n2, _ := strconv.Atoi(nums[1])
		ranges = append(ranges, [2]int{n1, n2})
	}
	sort.Slice(ranges, func(i, j int) bool {
		return ranges[i][0] < ranges[j][0]
	})
	merged := [][2]int{}
	tmp := ranges[0]
	for i := 1; i < len(ranges); i++ {
		if tmp[1]+1 >= ranges[i][0] {
			if ranges[i][1] > tmp[1] {
				tmp[1] = ranges[i][1]
			}
		} else {
			merged = append(merged, tmp)
			tmp = ranges[i]
		}
	}
	merged = append(merged, tmp)
	n := 0
	for i := 0; i < len(merged)-1; i++ {
		n += merged[i+1][0] - merged[i][1] - 1
	}
	n += 4294967295 - merged[len(merged)-1][1]
	return merged[0][1] + 1, n
}

File: 2016/20/solution.ts
import * as fs from "fs";
import * as path from "path";

const puzzleInput = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map((a) => a.split("-").map(Number)) as [number, number][];

const MAXIMUMIP = 4294967295;

class RangeCollection {
  ranges: [number, number][];
  constructor(ranges?: [number, number][]) {
    if (ranges && ranges.length) {
      this.ranges = ranges.sort((a, b) => a[0] - b[0]);
      this.addRange(ranges[0]);
    } else {
      this.ranges = [];
    }
  }

  addRange(newRange: [number, number]) {
    const ranges: [number, number][] = [];
    let curr = newRange;
    for (const range of this.ranges) {
      if (range[1] < curr[0] - 1) {
        ranges.push(range);
      } else if (curr[1] + 1 < range[0]) {
        ranges.push(curr);
        curr = range;
      } else {
        curr[0] = Math.min(curr[0], range[0]);
        curr[1] = Math.max(curr[1], range[1]);
      }
    }
    ranges.push(curr);
    this.ranges = ranges;
  }

  get lowestNonBlockedIP(): number {
    if (this.ranges.length === 0) return 0;
    return this.ranges[0][1] + 1;
  }

  get nNonBlockedIPs(): number {
    if (this.ranges.length === 0) return MAXIMUMIP + 1;
    let res = 0;
    let curr = -1;
    for (const range of this.ranges) {
      res += range[0] - curr - 1;
      curr = range[1];
    }
    return res + MAXIMUMIP - this.ranges[this.ranges.length - 1][1];
  }
}

const ranges = new RangeCollection(puzzleInput);

console.log(ranges.lowestNonBlockedIP);
console.log(ranges.nNonBlockedIPs);
